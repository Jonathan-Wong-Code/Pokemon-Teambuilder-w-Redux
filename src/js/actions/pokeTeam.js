import { ADD_TEAM, DELETE_TEAM, EDIT_TEAM, SET_SAVED_TEAMS } from './types';
import database from './../firebase/firebase';
import regeneratorRuntime from "regenerator-runtime";

export const addPokeTeam = (pokeTeam) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).push(pokeTeam);
  const newTeamObject = ({
    ...pokeTeam,
    id: response.key
  })

  dispatch({
    type : ADD_TEAM,
    team : newTeamObject
  });

  return response.key
};

export const deletePokeTeam = (id) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  database.ref(`users/${uid}/pokemon/${id}`).remove();

  dispatch({
    type : DELETE_TEAM,
    id
  });
};

export const editPokeTeam = (pokeTeam) => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  database.ref(`users/${uid}/pokemon/${pokeTeam.id}`).update(pokeTeam);

  dispatch({
    type : EDIT_TEAM,
    team : pokeTeam
  });
};

export const setSavedTeams = () => async (dispatch, getState) => {
  const uid = getState().auth.uid;
  const response = await database.ref(`users/${uid}/pokemon`).once('value');
  const savedTeams = {};
  response.forEach(team => {
    savedTeams[team.key] = {
      ...team.val(),
      id : team.key
    }
  });

  dispatch({
    type: SET_SAVED_TEAMS,
    savedTeams
  });
}
