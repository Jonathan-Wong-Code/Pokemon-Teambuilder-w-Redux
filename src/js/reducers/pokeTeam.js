import { ADD_TEAM, DELETE_TEAM, EDIT_TEAM, SET_SAVED_TEAMS, LOGOUT } from './../actions/types';

const pokeTeamReducer = (state = {}, action) => {
  switch(action.type) {

    case ADD_TEAM :
    return {...state, [action.team.id] : action.team }

    case DELETE_TEAM :
    const newState = Object.assign({}, state);
    delete newState[action.id];
    return newState;

    case EDIT_TEAM :
    return {...state, [action.team.id] : action.team}

    case SET_SAVED_TEAMS:
    return action.savedTeams;

    case LOGOUT:
    return {}

    default:
    return state;
  }
}

export default pokeTeamReducer;
