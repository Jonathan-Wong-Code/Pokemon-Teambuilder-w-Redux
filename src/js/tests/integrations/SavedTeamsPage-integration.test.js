import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import moment from 'moment';
import ConnectedSavedTeamsPage from '../../components/SavedTeamsPage';
import TestStore from '../testStore';
import pokeTeam from '../fixtures/pokeTeam';
import regeneratorRuntime from 'regenerator-runtime';

describe("The connected Saved Teams Page with pokeTeams saved", () => {
  let wrapper, initialState;
  beforeEach(() => {
    initialState = {
      savedTeams : pokeTeam
    };

    wrapper = mount(
      <TestStore initialState={initialState}>
        <MemoryRouter>
          <ConnectedSavedTeamsPage />
        </MemoryRouter>
      </TestStore> 
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('Should render a snapshot of the component', () => {
    expect(wrapper.find(`[data-test='saved-teams-component']`)).toMatchSnapshot();
  });

  it('Should render the Saved Teams Page component', () => {
    expect(wrapper.find(`[data-test='saved-teams-component']`).length).toBe(1);
  });

  //Saved Teams Page List

  it('Should render the saved teams list component', () => {
    expect(wrapper.find(`[data-test='saved-teams-list']`).length).toBe(1);
  });

  //SavedTeamsCell
  it('Should render the correct number of poke team cells', () => {
    expect(wrapper.find(`[data-test='saved-teams-cell']`).length).toBe(3);
  });

  it('Should render the Saved Teams Cell Component', () => {
    expect(wrapper.find(`[data-test='saved-teams-cell']`).length).toBe(3);
  });

  it('Should render the correct pokemon images', () => {
    expect(wrapper.find(`[data-test='saved-team-cell-img']`).at(0).prop('src')).toBe(pokeTeam.team1.pokemon[0].sprites.front_default);

    expect(wrapper.find(`[data-test='saved-team-cell-img']`).at(1).prop('src')).toBe(pokeTeam.team1.pokemon[1].sprites.front_default);

    expect(wrapper.find(`[data-test='saved-team-cell-img']`).at(2).prop('src')).toBe(pokeTeam.team1.pokemon[2].sprites.front_default);

    expect(wrapper.find(`[data-test='saved-team-cell-img']`).at(3).prop('src')).toBe(pokeTeam.team1.pokemon[3].sprites.front_default);

    expect(wrapper.find(`[data-test='saved-team-cell-img']`).at(4).prop('src')).toBe(pokeTeam.team1.pokemon[4].sprites.front_default);

    expect(wrapper.find(`[data-test='saved-team-cell-img']`).at(5).prop('src')).toBe(pokeTeam.team1.pokemon[5].sprites.front_default);
  });

  it('Should render the correct pokemon team name', () => {
    expect(wrapper.find(`[data-test='team-cell-name']`).at(0).text()).toBe(pokeTeam.team1.name);
  });

  it('Should render the correct pokemon team creation date', () => {
    expect(wrapper.find(`[data-test='team-cell-date']`).at(0).text()).toContain(moment(pokeTeam.team1.createdAt).format('MMM Do YYYY'));
  });
  
  it('Should render the correct pokemon team description', () => {
    expect(wrapper.find(`[data-test='team-cell-description']`).at(0).text()).toBe(pokeTeam.team1.description);
  });

  it('Should have a link that takes the user to the edit page', () => {
    expect(wrapper.find(`[data-test='team-cell-link']`).at(0).prop('to')).toBe(`/edit/${pokeTeam.team1.id}`);
  });
});

describe("The connected Saved Teams Page with no pokeTeams saved", () => {
  let wrapper, initialState;
  beforeEach(() => {
    initialState = {
      savedTeams : {}
    };

    wrapper = mount(
      <TestStore initialState={initialState}>
        <MemoryRouter>
          <ConnectedSavedTeamsPage />
        </MemoryRouter>
      </TestStore> 
    );
  });

  afterEach(() => {
    wrapper.unmount();
  });

  it('Should not render the Saved Teams List and render the error', () => {
    expect(wrapper.find(`[data-test='saved-teams-list']`).length).toBe(0);
    expect(wrapper.find(`[data-test='saved-teams-error']`).length).toBe(1);
  });
});
