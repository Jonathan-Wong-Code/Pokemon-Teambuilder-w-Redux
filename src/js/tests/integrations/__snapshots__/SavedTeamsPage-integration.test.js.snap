// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The connected Saved Teams Page with pokeTeams saved Should render a snapshot of the component 1`] = `
<section
  className="saved-teams-section"
  data-test="saved-teams-component"
>
  <h2
    className="saved-teams-section__heading"
  >
    My Saved Teams
  </h2>
  <SavedTeamsList
    savedTeams={
      Array [
        Object {
          "description": "This is pokeTeam1",
          "id": "team1",
          "name": "pokeTeam1",
          "pokemon": Array [
            Object {
              "name": "pokemon1",
              "sprites": Object {
                "front_default": "sprite1",
              },
            },
            Object {
              "name": "pokemon2",
              "sprites": Object {
                "front_default": "sprite2",
              },
            },
            Object {
              "name": "pokemon3",
              "sprites": Object {
                "front_default": "sprite3",
              },
            },
            Object {
              "name": "pokemon4",
              "sprites": Object {
                "front_default": "sprite4",
              },
            },
            Object {
              "name": "pokemon5",
              "sprites": Object {
                "front_default": "sprite5",
              },
            },
            Object {
              "name": "pokemon6",
              "sprites": Object {
                "front_default": "sprite6",
              },
            },
          ],
        },
        Object {
          "description": "This is pokeTeam2",
          "id": "team2",
          "name": "pokeTeam2",
          "pokemon": Array [
            Object {
              "name": "pokemon1",
              "sprites": Object {
                "front_default": "sprite1",
              },
            },
            Object {
              "name": "pokemon2",
              "sprites": Object {
                "front_default": "sprite2",
              },
            },
            Object {
              "name": "pokemon3",
              "sprites": Object {
                "front_default": "sprite3",
              },
            },
            Object {
              "name": "pokemon4",
              "sprites": Object {
                "front_default": "sprite4",
              },
            },
            Object {
              "name": "pokemon5",
              "sprites": Object {
                "front_default": "sprite5",
              },
            },
            Object {
              "name": "pokemon6",
              "sprites": Object {
                "front_default": "sprite6",
              },
            },
          ],
        },
        Object {
          "description": "This is pokeTeam3",
          "id": "team3",
          "name": "pokeTeam3",
          "pokemon": Array [
            Object {
              "name": "pokemon1",
              "sprites": Object {
                "front_default": "sprite1",
              },
            },
            Object {
              "name": "pokemon2",
              "sprites": Object {
                "front_default": "sprite2",
              },
            },
            Object {
              "name": "pokemon3",
              "sprites": Object {
                "front_default": "sprite3",
              },
            },
            Object {
              "name": "pokemon4",
              "sprites": Object {
                "front_default": "sprite4",
              },
            },
            Object {
              "name": "pokemon5",
              "sprites": Object {
                "front_default": "sprite5",
              },
            },
            Object {
              "name": "pokemon6",
              "sprites": Object {
                "front_default": "sprite6",
              },
            },
          ],
        },
      ]
    }
  >
    <div
      className="saved-teams"
      data-test="saved-teams-list"
    >
      <ul
        className="saved-teams__list"
      >
        <Connect(SavedTeamsCell)
          key="team1"
          team={
            Object {
              "description": "This is pokeTeam1",
              "id": "team1",
              "name": "pokeTeam1",
              "pokemon": Array [
                Object {
                  "name": "pokemon1",
                  "sprites": Object {
                    "front_default": "sprite1",
                  },
                },
                Object {
                  "name": "pokemon2",
                  "sprites": Object {
                    "front_default": "sprite2",
                  },
                },
                Object {
                  "name": "pokemon3",
                  "sprites": Object {
                    "front_default": "sprite3",
                  },
                },
                Object {
                  "name": "pokemon4",
                  "sprites": Object {
                    "front_default": "sprite4",
                  },
                },
                Object {
                  "name": "pokemon5",
                  "sprites": Object {
                    "front_default": "sprite5",
                  },
                },
                Object {
                  "name": "pokemon6",
                  "sprites": Object {
                    "front_default": "sprite6",
                  },
                },
              ],
            }
          }
        >
          <SavedTeamsCell
            deletePokeTeam={[Function]}
            team={
              Object {
                "description": "This is pokeTeam1",
                "id": "team1",
                "name": "pokeTeam1",
                "pokemon": Array [
                  Object {
                    "name": "pokemon1",
                    "sprites": Object {
                      "front_default": "sprite1",
                    },
                  },
                  Object {
                    "name": "pokemon2",
                    "sprites": Object {
                      "front_default": "sprite2",
                    },
                  },
                  Object {
                    "name": "pokemon3",
                    "sprites": Object {
                      "front_default": "sprite3",
                    },
                  },
                  Object {
                    "name": "pokemon4",
                    "sprites": Object {
                      "front_default": "sprite4",
                    },
                  },
                  Object {
                    "name": "pokemon5",
                    "sprites": Object {
                      "front_default": "sprite5",
                    },
                  },
                  Object {
                    "name": "pokemon6",
                    "sprites": Object {
                      "front_default": "sprite6",
                    },
                  },
                ],
              }
            }
          >
            <li
              className="team-cell"
              data-test="saved-teams-cell"
            >
              <ul
                className="team-cell__img-list"
              >
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon1"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite1"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon2"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite2"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon3"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite3"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon4"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite4"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon5"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite5"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon6"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite6"
                  />
                </li>
              </ul>
              <h2
                className="team-cell__heading"
                data-test="team-cell-name"
              >
                pokeTeam1
              </h2>
              <p
                className="team-cell__createdAt"
                data-test="team-cell-date"
              >
                Created: 
                Dec 31st 1969
              </p>
              <p
                className="team-cell__description"
                data-test="team-cell-description"
              >
                This is pokeTeam1
              </p>
              <div
                className="team-cell__buttons"
              >
                <Link
                  className="btn team-cell__button"
                  data-test="team-cell-link"
                  replace={false}
                  to="/edit/team1"
                >
                  <a
                    className="btn team-cell__button"
                    data-test="team-cell-link"
                    href="/edit/team1"
                    onClick={[Function]}
                  >
                    Edit
                  </a>
                </Link>
                <button
                  className="btn team-cell__button"
                  data-test="team-cell-delete"
                  onClick={[Function]}
                >
                  Delete
                </button>
              </div>
            </li>
          </SavedTeamsCell>
        </Connect(SavedTeamsCell)>
        <Connect(SavedTeamsCell)
          key="team2"
          team={
            Object {
              "description": "This is pokeTeam2",
              "id": "team2",
              "name": "pokeTeam2",
              "pokemon": Array [
                Object {
                  "name": "pokemon1",
                  "sprites": Object {
                    "front_default": "sprite1",
                  },
                },
                Object {
                  "name": "pokemon2",
                  "sprites": Object {
                    "front_default": "sprite2",
                  },
                },
                Object {
                  "name": "pokemon3",
                  "sprites": Object {
                    "front_default": "sprite3",
                  },
                },
                Object {
                  "name": "pokemon4",
                  "sprites": Object {
                    "front_default": "sprite4",
                  },
                },
                Object {
                  "name": "pokemon5",
                  "sprites": Object {
                    "front_default": "sprite5",
                  },
                },
                Object {
                  "name": "pokemon6",
                  "sprites": Object {
                    "front_default": "sprite6",
                  },
                },
              ],
            }
          }
        >
          <SavedTeamsCell
            deletePokeTeam={[Function]}
            team={
              Object {
                "description": "This is pokeTeam2",
                "id": "team2",
                "name": "pokeTeam2",
                "pokemon": Array [
                  Object {
                    "name": "pokemon1",
                    "sprites": Object {
                      "front_default": "sprite1",
                    },
                  },
                  Object {
                    "name": "pokemon2",
                    "sprites": Object {
                      "front_default": "sprite2",
                    },
                  },
                  Object {
                    "name": "pokemon3",
                    "sprites": Object {
                      "front_default": "sprite3",
                    },
                  },
                  Object {
                    "name": "pokemon4",
                    "sprites": Object {
                      "front_default": "sprite4",
                    },
                  },
                  Object {
                    "name": "pokemon5",
                    "sprites": Object {
                      "front_default": "sprite5",
                    },
                  },
                  Object {
                    "name": "pokemon6",
                    "sprites": Object {
                      "front_default": "sprite6",
                    },
                  },
                ],
              }
            }
          >
            <li
              className="team-cell"
              data-test="saved-teams-cell"
            >
              <ul
                className="team-cell__img-list"
              >
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon1"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite1"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon2"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite2"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon3"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite3"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon4"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite4"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon5"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite5"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon6"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite6"
                  />
                </li>
              </ul>
              <h2
                className="team-cell__heading"
                data-test="team-cell-name"
              >
                pokeTeam2
              </h2>
              <p
                className="team-cell__createdAt"
                data-test="team-cell-date"
              >
                Created: 
                Dec 31st 1969
              </p>
              <p
                className="team-cell__description"
                data-test="team-cell-description"
              >
                This is pokeTeam2
              </p>
              <div
                className="team-cell__buttons"
              >
                <Link
                  className="btn team-cell__button"
                  data-test="team-cell-link"
                  replace={false}
                  to="/edit/team2"
                >
                  <a
                    className="btn team-cell__button"
                    data-test="team-cell-link"
                    href="/edit/team2"
                    onClick={[Function]}
                  >
                    Edit
                  </a>
                </Link>
                <button
                  className="btn team-cell__button"
                  data-test="team-cell-delete"
                  onClick={[Function]}
                >
                  Delete
                </button>
              </div>
            </li>
          </SavedTeamsCell>
        </Connect(SavedTeamsCell)>
        <Connect(SavedTeamsCell)
          key="team3"
          team={
            Object {
              "description": "This is pokeTeam3",
              "id": "team3",
              "name": "pokeTeam3",
              "pokemon": Array [
                Object {
                  "name": "pokemon1",
                  "sprites": Object {
                    "front_default": "sprite1",
                  },
                },
                Object {
                  "name": "pokemon2",
                  "sprites": Object {
                    "front_default": "sprite2",
                  },
                },
                Object {
                  "name": "pokemon3",
                  "sprites": Object {
                    "front_default": "sprite3",
                  },
                },
                Object {
                  "name": "pokemon4",
                  "sprites": Object {
                    "front_default": "sprite4",
                  },
                },
                Object {
                  "name": "pokemon5",
                  "sprites": Object {
                    "front_default": "sprite5",
                  },
                },
                Object {
                  "name": "pokemon6",
                  "sprites": Object {
                    "front_default": "sprite6",
                  },
                },
              ],
            }
          }
        >
          <SavedTeamsCell
            deletePokeTeam={[Function]}
            team={
              Object {
                "description": "This is pokeTeam3",
                "id": "team3",
                "name": "pokeTeam3",
                "pokemon": Array [
                  Object {
                    "name": "pokemon1",
                    "sprites": Object {
                      "front_default": "sprite1",
                    },
                  },
                  Object {
                    "name": "pokemon2",
                    "sprites": Object {
                      "front_default": "sprite2",
                    },
                  },
                  Object {
                    "name": "pokemon3",
                    "sprites": Object {
                      "front_default": "sprite3",
                    },
                  },
                  Object {
                    "name": "pokemon4",
                    "sprites": Object {
                      "front_default": "sprite4",
                    },
                  },
                  Object {
                    "name": "pokemon5",
                    "sprites": Object {
                      "front_default": "sprite5",
                    },
                  },
                  Object {
                    "name": "pokemon6",
                    "sprites": Object {
                      "front_default": "sprite6",
                    },
                  },
                ],
              }
            }
          >
            <li
              className="team-cell"
              data-test="saved-teams-cell"
            >
              <ul
                className="team-cell__img-list"
              >
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon1"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite1"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon2"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite2"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon3"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite3"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon4"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite4"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon5"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite5"
                  />
                </li>
                <li
                  className="team-cell__img-item"
                >
                  <img
                    alt="pokemon6"
                    className="team-cell__img"
                    data-test="saved-team-cell-img"
                    src="sprite6"
                  />
                </li>
              </ul>
              <h2
                className="team-cell__heading"
                data-test="team-cell-name"
              >
                pokeTeam3
              </h2>
              <p
                className="team-cell__createdAt"
                data-test="team-cell-date"
              >
                Created: 
                Dec 31st 1969
              </p>
              <p
                className="team-cell__description"
                data-test="team-cell-description"
              >
                This is pokeTeam3
              </p>
              <div
                className="team-cell__buttons"
              >
                <Link
                  className="btn team-cell__button"
                  data-test="team-cell-link"
                  replace={false}
                  to="/edit/team3"
                >
                  <a
                    className="btn team-cell__button"
                    data-test="team-cell-link"
                    href="/edit/team3"
                    onClick={[Function]}
                  >
                    Edit
                  </a>
                </Link>
                <button
                  className="btn team-cell__button"
                  data-test="team-cell-delete"
                  onClick={[Function]}
                >
                  Delete
                </button>
              </div>
            </li>
          </SavedTeamsCell>
        </Connect(SavedTeamsCell)>
      </ul>
    </div>
  </SavedTeamsList>
</section>
`;
